src/main/java/org/xbill/DNS/ExtendedErrorCodeOption.java

```diff
@@ -0,0 +1,133 @@
+  public static final int NETWORK_ERROR = 23;
+  public static final int INVALID_DATA = 24;
+
+  @Getter private int errorCode;
+  @Getter private String text;
+
+  private static final Mnemonic codes =
+      new Mnemonic("EDNS Extended Error Codes", Mnemonic.CASE_SENSITIVE);
+
+  static {
+    codes.setMaximum(0xFFFF);
+    codes.setPrefix("EDE");
+    codes.add(OTHER, "Other");
+    codes.add(UNSUPPORTED_DNSKEY_ALGORITHM, "Unsupported DNSKEY Algorithm");
+    codes.add(UNSUPPORTED_DS_DIGEST_TYPE, "Unsupported DS Digest Type");
+    codes.add(STALE_ANSWER, "Stale Answer");
+    codes.add(FORGED_ANSWER, "Forged Answer");
+    codes.add(DNSSEC_INDETERMINATE, "DNSSEC Indeterminate");
+    codes.add(DNSSEC_BOGUS, "DNSSEC Bogus");
+    codes.add(SIGNATURE_EXPIRED, "Signature Expired");
+    codes.add(SIGNATURE_NOT_YET_VALID, "Signature Not Yet Valid");
+    codes.add(DNSKEY_MISSING, "DNSKEY Missing");
+    codes.add(RRSIGS_MISSING, "RRSIGs Missing");
+    codes.add(NO_ZONE_KEY_BIT_SET, "No Zone Key Bit Set");
+    codes.add(NSEC_MISSING, "NSEC Missing");
+    codes.add(CACHED_ERROR, "Cached Error");
+    codes.add(NOT_READY, "Not Ready");
+    codes.add(BLOCKED, "Blocked");
+    codes.add(CENSORED, "Censored");
+    codes.add(FILTERED, "Filtered");
+    codes.add(PROHIBITED, "Prohibited");
+    codes.add(STALE_NXDOMAIN_ANSWER, "Stale NXDOMAIN Answer");
+    codes.add(NOT_AUTHORITATIVE, "Not Authoritative");
+    codes.add(NOT_SUPPORTED, "Not Supported");
+    codes.add(NO_REACHABLE_AUTHORITY, "No Reachable Authority");
+    codes.add(NETWORK_ERROR, "Network Error");
+    codes.add(INVALID_DATA, "Invalid Data");
+  }
+
+  /** Creates an extended error code EDNS option. */
+  ExtendedErrorCodeOption() {
+    super(Code.EDNS_EXTENDED_ERROR);
+  }
+
+  /**
+   * Creates an extended error code EDNS option.
+   *
+   * @param errorCode the extended error.
+   * @param text optional error message intended for human readers.
+   */
+  public ExtendedErrorCodeOption(int errorCode, String text) {
+    super(Code.EDNS_EXTENDED_ERROR);
+    this.errorCode = errorCode;
+    this.text = text;
+  }
+
+  /**
+   * Creates an extended error code EDNS option.
+   *
+   * @param errorCode the extended error.
+   */
+  public ExtendedErrorCodeOption(int errorCode) {
+    this(errorCode, null);
+  }
+
+  @Override
+  void optionFromWire(DNSInput in) throws IOException {
+    errorCode = in.readU16();
+    if (in.remaining() > 0) {
+      byte[] data = in.readByteArray();
+      int len = data.length;
+
+      // EDE text may be null terminated but MUST NOT be assumed to be
+      if (data[data.length - 1] == 0) {
+        len--;
+      }
+
+      text = new String(data, 0, len, StandardCharsets.UTF_8);
+    }
+  }
+
+  @Override
+  void optionToWire(DNSOutput out) {
+    out.writeU16(errorCode);
+    if (text != null && text.length() > 0) {
+      out.writeByteArray(text.getBytes(StandardCharsets.UTF_8));
+    }
+  }
+
+  @Override
+  String optionToString() {
```

Comment:
Can we simply call this method `toString()`?

Reply:
No, it overrides a method (which is already public API). The base class already has `toString()` overridden:
https://github.com/dnsjava/dnsjava/blob/892fab2d826e6561c7c28b69461f3a965f2b397c/src/main/java/org/xbill/DNS/EDNSOption.java#L119-L122

Which e.g. produces `{Extended_DNS_Error: DNSSEC Indeterminate: not signed}`. I could still change the formatting of `optionToString()` and e.g. `Extended_DNS_Error` to `EDE` if that would be preferred.